examples/end2end_tree_reconstruction_test.py

```diff
@@ -0,0 +1,356 @@
+                for k, v in frames.items()
+            },
+            fossils=kwargs["fossil_interval"] is not None,
+            teeplot_dpi=100,
+            teeplot_outattrs=kwargs,
+            teeplot_outdir="/tmp",
+        )
+
+
+def test_reconstruct_one(
+    differentia_bitwidth: int,
+    surface_size: int,
+    fossil_interval: typing.Optional[int],
+    *,
+    visualize: bool,
+) -> typing.Dict[str, typing.Union[int, float, None]]:
+    """Test the reconstruction of a single phylogeny."""
+    print("=" * 80)
+    print(f"surface_size: {surface_size}")
+    print(f"differentia_bitwidth: {differentia_bitwidth}")
+    print(f"fossil_interval: {fossil_interval}")
+
+    frames = sample_reference_and_reconstruction(
+        differentia_bitwidth,
+        surface_size,
+        fossil_interval,
+    )
+
+    visualize_reconstruction(
+        frames,
+        differentia_bitwidth=differentia_bitwidth,
+        surface_size=surface_size,
+        fossil_interval=fossil_interval,
+        visualize=visualize,
+    )
+    reconstruction_error = alifestd_calc_triplet_distance_asexual(
+        alifestd_collapse_unifurcations(frames["true"]), frames["reconst"]
+    )
+
+    reconstruction_error_dropped_fossils = (
+        alifestd_calc_triplet_distance_asexual(
+            alifestd_collapse_unifurcations(frames["true_dropped_fossils"]),
+            frames["reconst_dropped_fossils"],
+        )
+    )
+
+    print(f"{reconstruction_error=}")
+    print(f"{reconstruction_error_dropped_fossils=}")
+    assert 0 <= reconstruction_error <= 1  # should be in the range [0,1]
+
+    return {
+        "differentia_bitwidth": differentia_bitwidth,
+        "surface_size": surface_size,
+        "fossil_interval": fossil_interval,
+        "error": reconstruction_error,
+        "error_dropped_fossils": reconstruction_error_dropped_fossils,
+    }
+
+
+def _parse_args():
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--visualize", action="store_true")
+    return parser.parse_args()
+
+
+if __name__ == "__main__":
+    sys.setrecursionlimit(100000)
+    args = _parse_args()
+    reconstruction_errors = pd.DataFrame(
+        [
+            test_reconstruct_one(
+                differentia_bitwidth,
+                surface_size,
+                fossil_interval,
+                visualize=args.visualize,
+            )
+            for (
+                fossil_interval,
+                surface_size,
+                differentia_bitwidth,
+            ) in product((None, 50, 200), (256, 64, 16), (64, 8, 1))
+        ]
+    ).sort_values(["surface_size", "differentia_bitwidth"], ascending=False)
+
+    reconstruction_errors.to_csv("/tmp/end2end-reconstruction-error.csv")
+
+    # error should increase with decreasing surface size
+    tolerance = 0.02
+    assert (  # type: ignore
+        reconstruction_errors.groupby("fossil_interval", dropna=False)["error"]
+        .agg(
```

Comment:
Instead of making the tolerances stack, does it work to just require each link to be individually smaller than tolerance?

Reply:
Can we also log how many out of how many results are non-increasing but within tolerance?